lib/std/debug.zig

```diff
@@ -224,24 +529,49 @@ pub fn assert(ok: bool) void {
 }
 
 pub fn panic(comptime format: []const u8, args: anytype) noreturn {
+    panicExtra(null, format, args);
+}
+
+/// `panicExtra` is useful when you want to print out an `@errorReturnTrace`
+/// and also print out some values.
+pub fn panicExtra(
+    trace: ?*builtin.StackTrace,
+    comptime format: []const u8,
+    args: anytype,
+) noreturn {
     @setCold(true);
-    // TODO: remove conditional once wasi / LLVM defines __builtin_return_address
-    const first_trace_addr = if (builtin.os.tag == .wasi) null else @returnAddress();
-    panicExtra(null, first_trace_addr, format, args);
+    const size = 0x1000;
+    const trunc_msg = "(msg truncated)";
+    var buf: [size + trunc_msg.len]u8 = undefined;
+    // a minor annoyance with this is that it will result in the NoSpaceLeft
+    // error being part of the @panic stack trace (but that error should
+    // only happen rarely)
+    const msg = std.fmt.bufPrint(buf[0..size], format, args) catch |err| switch (err) {
+        std.fmt.BufPrintError.NoSpaceLeft => blk: {
+            std.mem.copy(u8, buf[size..], trunc_msg);
+            break :blk &buf;
+        },
+    };
+    builtin.panic(msg, trace);
 }
```

Comment:
A buffer of size 4096 is used for formating. If we go over, we append "(msg truncated)" but otherwise swallow it.
We use `builtin.panic` instead of `@panic` because we want to pass in a trace. Also, if we used `@panic` then we would get a trace from bufPrint error!